optimum/habana/sentence_transformers/st_gaudi_trainer.py

```diff
@@ -317,13 +323,14 @@ def compute_loss(
         # Insert the wrapped (e.g. distributed or compiled) model into the loss function,
         # if the loss stores the model. Only called once per process
         # from https://github.com/UKPLab/sentence-transformers/blob/v3.1.0/sentence_transformers/trainer.py#L337
+        # Override the model in the loss function with the original model to prevent recursion
         if (
             model == self.model_wrapped
             and model != self.model  # Only if the model is wrapped
             and hasattr(loss_fn, "model")  # Only if the loss stores the model
-            and loss_fn.model != model  # Only if the wrapped model is not already stored
+            and loss_fn.model != self.model  # Assign the original model instead
```

Comment:
@kplau1128 this does not sound correct to me! Here the goal is to insert the wrapped model (distributed or compiled) into the loss function, not the original model

Reply:
I see that without this fix the recursion happen, but the workaround is different form the intention

Reply:
Here it does not do anything since `loss_fn.model is self.model`

Reply:
You are right.

The crux of the issue lies in circular references introduced by the combined wrappers. I will rework the workaround only to modify the `override_model_in_loss` function to skip override loss when combined wrappers.